pkg/autoscaler/stats_scraper.go

```diff
@@ -117,41 +121,60 @@ func newServiceScraperWithClient(
+func (s *ServiceScraper) Scrape(statsCh chan<- *StatMessage) {
 	readyPodsCount, err := resources.FetchReadyAddressCount(s.endpointsLister, s.namespace, s.scrapeTargetService)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to get endpoints")
+		s.logger.Errorf("Failed to get endpoints: %v", err)
+		return
 	}
 
 	if readyPodsCount == 0 {
-		return nil, nil
+		s.logger.Debug("Do not scrape stats because no ready pods.")
+		return
 	}
 
-	stat, err := s.sClient.Scrape(s.url)
-	if err != nil {
-		return nil, err
+	sampleSize := populationMeanSampleSize(readyPodsCount)
+	if sampleSize < 1 {
+		s.logger.Debug("Do not scrape stats because sample size is less than 1.")
+		return
 	}
-
-	// Assumptions:
-	// 1. Traffic is routed to pods evenly.
-	// 2. A particular pod can stand for other pods, i.e. other pods have
-	//    similar concurrency and QPS.
-	//
-	// Hide the actual pods behind scraper and send only one stat for all the
-	// customer pods per scraping. The pod name is set to a unique value, i.e.
-	// scraperPodName so in autoscaler all stats are either from activator or
-	// scraper.
-	extrapolatedStat := Stat{
-		Time:                             stat.Time,
-		PodName:                          scraperPodName,
-		AverageConcurrentRequests:        stat.AverageConcurrentRequests * float64(readyPodsCount),
-		AverageProxiedConcurrentRequests: stat.AverageProxiedConcurrentRequests * float64(readyPodsCount),
-		RequestCount:                     stat.RequestCount * int32(readyPodsCount),
-		ProxiedRequestCount:              stat.ProxiedRequestCount * int32(readyPodsCount),
+	// Sleep time during each HTTP call to the metric service.
+	sleepTime := scrapeTickInterval / time.Duration(sampleSize)
+
+	for i := 0; i < sampleSize; i++ {
+		go func() {
+			stat, err := s.sClient.Scrape(s.url)
+			if err != nil {
+				return
+			}
+
+			// Assumptions:
+			// 1. Traffic is routed to pods evenly.
+			// 2. A particular pod can stand for other pods, i.e. other pods have
+			//    similar concurrency and QPS.
+			//
+			// Hide the actual pods behind scraper and send only one stat for all the
+			// customer pods per scraping. The pod name is set to a unique value, i.e.
+			// scraperPodName so in autoscaler all stats are either from activator or
+			// scraper.
+			now := time.Now()
+			extrapolatedStat := Stat{
+				Time:                             &now,
+				PodName:                          scraperPodName,
+				AverageConcurrentRequests:        stat.AverageConcurrentRequests * float64(readyPodsCount),
+				AverageProxiedConcurrentRequests: stat.AverageProxiedConcurrentRequests * float64(readyPodsCount),
+				RequestCount:                     stat.RequestCount * int32(readyPodsCount),
+				ProxiedRequestCount:              stat.ProxiedRequestCount * int32(readyPodsCount),
+			}
+
+			statsCh <- &StatMessage{
+				Stat: extrapolatedStat,
+				Key:  s.metricKey,
+			}
+		}()
+
+		if i != sampleSize-1 {
+			time.Sleep(sleepTime)
+		}
```

Comment:
I think I'm having second thoughts on the implementation in general. I know I've been the one advocating to wrap multiple scrapes into the "Scrape" call and report them as one value.

However, maybe it's a much simpler approach (and kinda already implemented anyway) to continuously adjust the scrape interval and run off a single timer.

Essentially the current code now runs off of two timers: The 1 second ticker and the adjusted sleep interval in here. These two timers can get skewed and you'll end up with multiple concurrent Scrape() calls on a single revision if for some reason the latencies of scraping increase.

Should we instead run on one timer per revision, which is continuously adjusted with the population and ditch the uniform 1 second ticker?

Reply:
1. I think we need another signal to tell "this is the time to change the scrape rate". Do you mean you prefer to use a counter instead of a ticker?
1. `collector` doesn't know the pods count. I don't want to introduce `Endpoints` informer to `collector` to make it complicated.
1. Why don't we allow multiple concurrent `Scrape()` calls on a single revision? `RoundTripper` must be safe for concurrent use by multiple goroutines.
1. In the case that for some reason the latencies of scraping increase, and we use a single ticker to call `Scrape()` in single goroutine, what's the behavior you expect if latency is higher than scrape rate?

Reply:
1/2. I think I had even more encapsulation in mind: The scraper would have a `StartScraping` method that spawns a goroutine and the whole thing decides how to behave itself. It wouldn't be controller by the outside at all.
3. We could also, as you said, just send all of the scrapes concurrently. That'd solve my concerns here as well.
4. You're right, we don't need to be aware of long latencies here as it'll self-adjust through channel backpressure.

On a meta level my concern is: We should only have one timer driving the scrape interval. It's either the outside one (1s) or the inside one (1s/population). Having two is imo hard to follow, especially if you want to treat each "internal" iteration as a separate Stat anyway.

Reply:
If only one timer, I prefer to the inside one. However I also prefer small PRs. So I just remove the inside timer and send the scrapes concurrently at this point.